firedrake/assemble.py

```diff
@@ -1980,28 +1962,28 @@ def _as_parloop_arg(tsfc_arg, self):
 @_as_parloop_arg.register(kernel_args.OutputKernelArg)
 def _as_parloop_arg_output(_, self):
     rank = len(self._form.arguments())
-    tensor = self._indexed_tensor
+    pyop2_tensor = _as_pyop2_tensor(self._tensor, self._indices)
```

Comment:
Deprecation of `_FormHandler.index_tensor` looks almost independent of what you are trying to achieve in this PR. Putting aside if `_FormHandler.index_tensor` is good or not, if we have to introduce a free function `_as_pyop2_tensor` as a result of the deprecation, we should probably just keep `_FormHandler.index_tensor` (and remove `form` and `diagonal` arguments).

Reply:
Arguably it is odd to have a `_FormHandler` method that doesn't actually reference a form at any point. Also I think that using `staticmethod`s in this way is just a worse way of doing things than having free functions (I'm allowed to say this because I wrote it originally!).

Reply:
I think this is no longer necessary to discuss given my other changes. Please resolve if you agree.